returnn/tf/engine.py

```diff
@@ -2316,27 +2320,80 @@ def extra_fetches_callback(seq_idx, seq_tag, **kwargs):
+      for output_layer_idx in range(num_output_layers):
+        outputs.append(kwargs["output_" + output_layer_names[output_layer_idx]])
+        beam_scores.append(kwargs["beam_scores_" + output_layer_names[output_layer_idx]])
+        if do_eval and target_keys[output_layer_idx] is not None:
+          targets.append(kwargs["target_" + target_keys[output_layer_idx]])
+        else:
+          targets.append(None)
 
       n_batch = len(seq_idx)  # without beam
       assert n_batch == len(seq_tag)
 
       # noinspection PyShadowingNames
-      for target_idx in range(num_targets):
-        if beam_scores[target_idx] is not None:
-          assert beam_scores[target_idx].shape == (n_batch, out_beam_sizes[target_idx])
+      for output_layer_idx in range(num_output_layers):
+        if beam_scores[output_layer_idx] is not None:
+          assert beam_scores[output_layer_idx].shape == (n_batch, out_beam_sizes[output_layer_idx])
 
-        assert n_batch * (out_beam_sizes[target_idx] or 1) == len(outputs[target_idx])
-        if do_eval and targets[target_idx] is not None:
-          assert n_batch == len(targets[target_idx])
+        assert n_batch * (out_beam_sizes[output_layer_idx] or 1) == len(outputs[output_layer_idx])
+        if do_eval and targets[output_layer_idx] is not None:
+          assert n_batch == len(targets[output_layer_idx])
 
-        if output_layers[target_idx].output.dim == 256 and output_layers[target_idx].output.sparse:
+        if output_layers[output_layer_idx].output.dim == 256 and output_layers[output_layer_idx].output.sparse:
           # Interpret output as bytes/utf8-string.
-          outputs[target_idx] = bytearray(outputs[target_idx]).decode("utf8")
+          outputs[output_layer_idx] = bytearray(outputs[output_layer_idx]).decode("utf8")
+
+      # Create lists with serialized data. All of length num_output_layers.
+      serialized_outputs = []  # type: typing.List[typing.Optional[typing.Union[str,numpy.ndarray]]]
+      serialized_targets = []  # type: typing.List[typing.Optional[typing.Union[str,numpy.ndarray]]]
+      # noinspection PyShadowingNames
+      for output_layer_idx in range(num_output_layers):
+        if output_layers[output_layer_idx].output.sparse:
+          # Try to get the vocab corresponding to the sparse dim
+          vocab = None
+          if output_layers[output_layer_idx].output.sparse_dim.vocab is not None:
+            vocab = output_layers[output_layer_idx].output.sparse_dim.vocab
+          elif target_keys[output_layer_idx] and (
+            self.network.extern_data.get_data(target_keys[output_layer_idx]).vocab is not None):
+            vocab = self.network.get_extern_data(target_keys[output_layer_idx]).vocab
+          elif target_keys[output_layer_idx] and target_keys[output_layer_idx] in dataset.labels:
+            vocab = Vocabulary.create_vocab_from_labels(dataset.labels[target_keys[output_layer_idx]])
```

Comment:
I think you can simplify the code.
```suggestion
          vocab = output_layers[output_layer_idx].output.sparse_dim.vocab
          if not vocab and target_keys[output_layer_idx]:
            vocab = self.network.get_extern_data(target_keys[output_layer_idx]).vocab
            if not vocab:
              vocab = Vocabulary.create_vocab_from_labels(dataset.labels[target_keys[output_layer_idx]])
```